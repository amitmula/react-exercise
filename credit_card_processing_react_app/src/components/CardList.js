import React, { Component } from "react";
import Table from 'react-bootstrap/Table'
import io from 'socket.io-client';

export default class CardList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            cards: [],
            socket: {}
        }
    }

    componentDidMount() {

        this.state.socket = io.connect(process.env.REACT_APP_API_ENDPOINT);

        this.state.socket.on('connect', () => {
            this.state.socket.emit('join channel', 'cardAdded', function (confirmation) {
                console.log("joined channel cardAdded -> ", confirmation);
            });
        });

        this.state.socket.on('connect_error', () => {
            console.error('Error connecting to notification service.')
        });

        this.state.socket.on('card added', (socketData) => {
            var respData = socketData;
            console.log('respData -> ', respData)
            if (respData) {
                var currentCards = this.state.cards
                console.log('currentCards -> ', currentCards)
                currentCards.push(respData)
                console.log('updatedCards -> ', currentCards)
                this.setState({...this.state, cards: currentCards})
            }
        });

        fetch(process.env.REACT_APP_API_ENDPOINT + process.env.REACT_APP_API_CONTEXT_PATH + process.env.REACT_APP_API_GET_ALL_CARDS_PATH, {
            method: 'GET',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json',
            }
        }).then(response => response.json())
            .catch(error => console.error(error))
            .then(resp => {
                this.setState({ ...this.state, cards: resp });
            })
    }

    render() {
        return (
            <Table striped bordered hover>
                <thead>
                    <tr>
                        <th>Name</th>
                        <th>Card Number</th>
                        <th>Balance</th>
                        <th>Limit</th>
                    </tr>
                </thead>
                <tbody>
                    {this.state.cards.length ? this.state.cards.map((card, index) => {
                        return (
                            <tr key={index}>
                                <td>{card.name}</td>
                                <td>{card.number}</td>
                                <td>{card.balance}</td>
                                <td>{card.limit}</td>
                            </tr>
                        )
                    }) : <tr><td colSpan="4" className="noResults">No cards found</td></tr>}
                </tbody>
            </Table>
        )
    }
}