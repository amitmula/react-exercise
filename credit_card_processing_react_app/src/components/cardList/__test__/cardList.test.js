import React from 'react'
import ReactDom from 'react-dom'
import CardList from '../cardList'

import { render, cleanup } from '@testing-library/react'
import "@testing-library/jest-dom/extend-expect"
import renderer from 'react-test-renderer'

afterEach(cleanup)

it("renders without crashing", () => {
    const div = document.createElement("div")
    ReactDom.render(<CardList />, div)
    ReactDom.unmountComponentAtNode(div)
})

it("renders cardList with props", () => {
    const { getByTestId } = render(<CardList testProp="testPropVal" />)
    expect(getByTestId('cardsListTable').getAttribute("test-attribute")).toBe('testPropVal');
})

it("renders cardList with props", () => {
    const { getByTestId } = render(<CardList testProp="some test val" />)
    expect(getByTestId('cardsListTable').getAttribute("test-attribute")).toBe('some test val');
})

it("matches snapshot", () => {
    const tree = renderer.create(<CardList testProp="testPropVal" />).toJSON()
    expect(tree).toMatchSnapshot()
})
