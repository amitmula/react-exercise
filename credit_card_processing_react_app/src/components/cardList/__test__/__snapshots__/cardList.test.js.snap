// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<table
  className="table table-striped table-bordered table-hover"
  data-testid="cardsListTable"
  test-attribute="testPropVal"
>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Card Number
      </th>
      <th>
        Balance
      </th>
      <th>
        Limit
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        className="noResults"
        colSpan="4"
      >
        No cards found
      </td>
    </tr>
  </tbody>
</table>
`;
